### PR TITLE
Assignment 1 submittion fix

### PR DESCRIPTION
**Pull request by olakalisz**

### CODE REVIEW

**Comment by flerdacodeu** [comment]
Perfect, thanks!

File: Assignment1_q1_alternative_sol.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from collections import Counter
+
+# Returns true if strings 's1' and 's2' are permutations of the other, false otherwise
```
Comment (flerdacodeu) [inline Assignment1_q1_alternative_sol.py]: This comments are better suited as docstring comments.
A docstring is a string that is placed as the first statement within a class, method, or function.
Python interprets them in a special way, e.g., the `help()` command will show the docstring for a function.

https://www.python.org/dev/peps/pep-0257/

File: Assignment1_q1_alternative_sol.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+def is_permutation(s1, s2):
+
+    # We are case-insensitive
+    s1 = s1.lower()
+    s2 = s2.lower()
+
+    # Character count in both strings needs to be the same for them to be permutations
+    return Counter(s1) == Counter(s2)
+
+# Test examples
```
Comment (flerdacodeu) [inline Assignment1_q1_alternative_sol.py]: Besides these examples, you should actually write automated tests for your code.
Automated tests (using the `unittest` module) should consider examples like the one below, but also focus on corner cases and error cases.

File: Assignment1_q1_permutations.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    # We are case-insensitive
+    s1 = s1.lower()
+    s2 = s2.lower()
+
+    # If strings lengths are not equal they cannot be permutations of the other
+    if len(s1) != len(s2):
+        return False
+
+    # Initialise two dictionaries 'lc1' and 'lc2' to count occurrences of letters in 's1' and 's2'
+    lc1 = {}    # Letter count for 's1' (character count)
```
Comment (flerdacodeu) [inline Assignment1_q1_permutations.py]: Consider more descriptive names.
For example `character_counts1` or `char_dict1`.

Comment (flerdacodeu) [inline Assignment1_q1_permutations.py]: Given the comment above, the one at the end of the line seems redundant.

File: Assignment1_q1_permutations.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            lc1[l1] = 1
+
+        if l2 in lc2:
+            lc2[l2] += 1
+        else:
+            lc2[l2] = 1
+
+    # Check whether 's1' and 's2' contain the same set of letters and if the number of occurrences of each
+    # letter in both strings is the same
+    for character in lc1:
```
Comment (flerdacodeu) [inline Assignment1_q1_permutations.py]: You are basically comparing if two dictionaries are the same: you can simply use equality for that.

File: Assignment1_q2_Node.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Node is a unit which is used in my implementation of Singly Linked List
+class Node(object):
```
Comment (flerdacodeu) [inline Assignment1_q2_Node.py]: Since `Node` is part of the linked list data structure, it could be defined in the same module (i.e., in the same file).

File: Assignment1_q2_Node.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+    def __init__(self, data = None, next_node = None):
+        # Node is storing some data and a pointer to the next node
+        self.data = data
```
Comment (flerdacodeu) [inline Assignment1_q2_Node.py]: Since these variables are private, you can name them `__data` and `__next` so that code outside this class cannot (accidentally) use them.

File: Assignment1_q2_Singly_Linked_List.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
+from Assignment1_q2_Node import *
```
Comment (flerdacodeu) [inline Assignment1_q2_Singly_Linked_List.py]: As said above, you want to probably put the `Node` class in this file, but (generally) avoid importing `*` into a module: this means that if someone defines new things in the other module, all of them will be available (and might affect how your code works). Instead, import only what you need.

File: Assignment1_q2_Singly_Linked_List.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    # (The idea is to have two pointers, both set to head at the start
+    # we move the first pointer k times - and thus it has len(list) - k
+    # elements till the end of the list - and then start moving both pointers
+    # until the first pointer reaches the end of the list, then return the second pointer)
+    def get_from_last(self, k):
+        main_pointer = self.head
+        helper_pointer = self.head
+
+        if self.head is None:
+            raise ValueError("Empty list!")
```
Comment (flerdacodeu) [inline Assignment1_q2_Singly_Linked_List.py]: `IndexError` might be more appropriate.

File: Assignment1_q2_Singly_Linked_List.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    def __init__(self, head = None):
+        self.head = head
+
+    # Inserting elements to the front of the list
+    def insert_first(self, data):
+        new_node = Node(data)
+        new_node.set_next_node(self.head)
+        self.head = new_node
+
+    # Get kth to the last element of the list - assignment question
```
Comment (flerdacodeu) [inline Assignment1_q2_Singly_Linked_List.py]: The function actually returns a `Node` rather than the element of the list.
From the user's perspective, they have a list of, e.g., strings and the want the k-th to last string, note the `Node`.

**Review by flerdacodeu — CHANGES_REQUESTED** [review]
A few suggestions for you.

If you need help with anything, please do let me know.

File: Assignment1_q1_alternative_sol.py (hunk 1 above)
Comment (olakalisz) [reply Assignment1_q1_alternative_sol.py]: Done!

File: Assignment1_q1_alternative_sol.py (hunk 2 above)
Comment (olakalisz) [reply Assignment1_q1_alternative_sol.py]: I have added test files :)

File: Assignment1_q1_permutations.py (hunk 3 above)
Comment (olakalisz) [reply Assignment1_q1_permutations.py]: Renamed!

Comment (olakalisz) [reply Assignment1_q1_permutations.py]: ... and deleted 👍

File: Assignment1_q1_permutations.py (hunk 4 above)
Comment (olakalisz) [reply Assignment1_q1_permutations.py]: using == now

File: Assignment1_q2_Node.py (hunk 5 above)
Comment (olakalisz) [reply Assignment1_q2_Node.py]: Done!

File: Assignment1_q2_Node.py (hunk 6 above)
Comment (olakalisz) [reply Assignment1_q2_Node.py]: I haven't seen the syntax of private variables in python before, thanks for this suggestion! I have changed it now

File: Assignment1_q2_Singly_Linked_List.py (hunk 7 above)
Comment (olakalisz) [reply Assignment1_q2_Singly_Linked_List.py]: I deleted it :) it makes a lot of sense I'll avoid it next time, thanks!

File: Assignment1_q2_Singly_Linked_List.py (hunk 9 above)
Comment (olakalisz) [reply Assignment1_q2_Singly_Linked_List.py]: Changed the comment

File: Assignment1_q2_Singly_Linked_List.py (hunk 8 above)
Comment (olakalisz) [reply Assignment1_q2_Singly_Linked_List.py]: it returns IndexError now but I left ValueError for k out of bounds, is that okay?

File: Assignment1_q1.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
+"""Checks if two strings are permutations of the other."""
```
Comment (flerdacodeu) [inline Assignment1_q1.py]: nit: "of each other"

File: Assignment1_q2.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+    def get_head(self):
+        return self.__head
+
+    def insert_first(self, data):
+        """Inserting elements to the front of the list."""
+        new_node = Node(data)
+        new_node.set_next_node(self.__head)
+        self.__head = new_node
+
```
Comment (flerdacodeu) [inline Assignment1_q2.py:32]: nit: You can remove one of the two blank lines here.

File: Assignment1_q2_test.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+import Assignment1_q2
+import unittest
+
+class Q2Test(unittest.TestCase):
+
+  def testEmptyCase(self):
+    list = Assignment1_q2.SinglyLinkedList()
+    with self.assertRaises(IndexError):
+        list.get_from_last(1)
```
Comment (flerdacodeu) [inline Assignment1_q2_test.py:11]: You might want to check that get_from_last(0) also fails (within this method).

File: Assignment1_q2_Singly_Linked_List.py (hunk 8 above)
Comment (flerdacodeu) [reply Assignment1_q2_Singly_Linked_List.py]: Yes, it is alright this way for the empty list as input.
One thing you can do is add to the docstring that method should only be called if the list in not empty.

**Review by flerdacodeu — APPROVED** [review]
Thanks Aleksandra.
Just minor suggestions left.
Well done!

Feel free to merge your change whenever you are happy with it.

File: Assignment1_q1.py (hunk 10 above)
Comment (olakalisz) [reply Assignment1_q1.py]: Changed!

File: Assignment1_q2.py (hunk 11 above)
Comment (olakalisz) [reply Assignment1_q2.py:32]: Done!

File: Assignment1_q2_test.py (hunk 12 above)
Comment (olakalisz) [reply Assignment1_q2_test.py:11]: I have added a new test case 👍

**Comment by olakalisz** [comment]
I have fixed the little bugs you mentioned :) I think it's ready to be merged :)
